src/assets/TokensController.ts

```diff
@@ -213,102 +260,237 @@ export class TokensController extends BaseController<
     decimals: number,
     image?: string,
   ): Promise<Token[]> {
+    const currentChainId = this.config.chainId;
     const releaseLock = await this.mutex.acquire();
     try {
       address = toChecksumHexAddress(address);
-      const { tokens, ignoredTokens } = this.state;
-      const isERC721 = await this._detectIsERC721(address);
-      const newEntry: Token = { address, symbol, decimals, image, isERC721 };
-      const previousEntry = tokens.find(
+      const { tokens, ignoredTokens, detectedTokens } = this.state;
+      const newTokens: Token[] = [...tokens];
+      const [isERC721, tokenMetadata] = await Promise.all([
+        this._detectIsERC721(address),
+        this.fetchTokenMetadata(address),
+      ]);
+      if (currentChainId !== this.config.chainId) {
+        throw new Error(
+          'TokensController Error: Switched networks while adding token',
+        );
+      }
+      const newEntry: Token = {
+        address,
+        symbol,
+        decimals,
+        image:
+          image ||
+          formatIconUrlWithProxy({
+            chainId: this.config.chainId,
+            tokenAddress: address,
+          }),
+        isERC721,
+        aggregators: formatAggregatorNames(tokenMetadata?.aggregators || []),
+      };
+      const previousEntry = newTokens.find(
         (token) => token.address.toLowerCase() === address.toLowerCase(),
       );
       if (previousEntry) {
-        const previousIndex = tokens.indexOf(previousEntry);
-        tokens[previousIndex] = newEntry;
+        const previousIndex = newTokens.indexOf(previousEntry);
+        newTokens[previousIndex] = newEntry;
       } else {
-        tokens.push(newEntry);
+        newTokens.push(newEntry);
       }
 
       const newIgnoredTokens = ignoredTokens.filter(
         (tokenAddress) => tokenAddress.toLowerCase() !== address.toLowerCase(),
       );
-      const { newAllTokens, newAllIgnoredTokens } = this._getNewAllTokensState(
-        tokens,
-        newIgnoredTokens,
+      const newDetectedTokens = detectedTokens.filter(
+        (token) => token.address.toLowerCase() !== address.toLowerCase(),
```

Comment:
This line is not covered. We don't have any tests that add a token that was previously a detected token.